ee/apps/omnichannel-transcript/src/OmnichannelTranscript.ts

```diff
@@ -1,16 +1,252 @@
+import type { Upload, Message, QueueWorker, Translation } from '../../../../apps/meteor/server/sdk';
+import type { Logger } from '../../../../apps/meteor/server/lib/logger/Logger';
+
+const isPromiseRejectedResult = (result: any): result is PromiseRejectedResult => result.status === 'rejected';
+
+type WorkDetails = {
+	rid: IRoom['_id'];
+	userId: IUser['_id'];
+};
+
+type WorkDetailsWithSource = WorkDetails & {
+	from: string;
+};
 
 export class OmnichannelTranscript extends ServiceClass implements IOmnichannelTranscriptService {
 	protected name = 'omnichannel-transcript';
 
-	constructor() {
-		super();
+	private worker: PdfWorker;
 
+	private log: Logger;
+
+	constructor(
+		private readonly uploadService: typeof Upload,
+		private readonly messageService: typeof Message,
+		private readonly queueService: typeof QueueWorker,
+		private readonly translationService: typeof Translation,
+		loggerClass: typeof Logger,
+	) {
+		super();
+		this.worker = new PdfWorker();
+		// eslint-disable-next-line new-cap
+		this.log = new loggerClass('OmnichannelTranscript');
 		// your stuff
 	}
 
-	getConfig(): unknown {
-		return null;
+	private getMessagesFromRoom({ rid }: { rid: string }): Promise<IMessage[]> {
+		return Messages.findLivechatMessages(rid, {
+			sort: { ts: 1 },
+			projection: { _id: 1, msg: 1, u: 1, t: 1, ts: 1, attachments: 1, files: 1 },
+		}).toArray();
+	}
+
+	async requestTranscript({ details }: { details: WorkDetails }): Promise<void> {
+		this.log.log(`Requesting transcript for room ${details.rid} by user ${details.userId}`);
+		const room = await LivechatRooms.findOneById(details.rid);
+		if (!room) {
+			throw new Error('room-not-found');
+		}
+
+		if (room.open) {
+			throw new Error('room-still-open');
+		}
+
+		if (!room.servedBy || !room.v) {
+			throw new Error('improper-room-state');
+		}
+
+		// Don't request a transcript if there's already one requested :)
+		if (room.pdfTranscriptRequested) {
+			// TODO: use logger
+			this.log.log(`Transcript already requested for room ${details.rid}`);
+			return;
+		}
+
+		await LivechatRooms.setTranscriptRequestedPdfById(details.rid);
+
+		// Even when processing is done "in-house", we still need to queue the work
+		// to avoid blocking the request
+		this.log.log(`Queuing work for room ${details.rid}`);
+		await this.queueService.queueWork('work', `${this.name}.workOnPdf`, {
+			template: 'omnichannel-transcript',
+			details: { userId: details.userId, rid: details.rid, from: this.name },
+		});
+	}
+
+	private async getFiles(
+		userId: string,
+		messages: IMessage[],
+	): Promise<(Pick<IMessage, '_id' | 'ts' | 'u' | 'msg'> & { files: ({ name?: string; buffer: Buffer | null } | undefined)[] })[]> {
+		return Promise.all(
+			messages.map(async (message: IMessage) => {
+				if (!message.attachments || !message.attachments.length) {
+					// If there's no attachment and no message, what was sent? lol
```

Comment:
AFAIK, there's a third case where UiKit blocks are sent, like for example when you interact with a chatbot that has quick reply buttons with messages. In this case, its data would be stored inside `blocks` property of message. [Here's](https://github.com/RocketChat/Apps.WhatsApp/blob/634aee83ec5f35add30bf5ff600e4f5f820cffe6/helpers/message.ts#L539) a parser for these block from one of the apps

Reply:
Good point, but I'm not sure either if UIKit blocks are part of the scope 🤔 
Let me see the parser and if the thing is not "too complex", i'll add it. Otherwise, I'll try to filter out those messages